### PR TITLE
[RFC] Codify CI policy

### PR DESCRIPTION
**Pull request by ppannuto**
### Pull Request Overview

This takes a first cut at writing down a structure and policy for CI.

### Testing Strategy

N/A.

### TODO or Help Wanted

Feedback!

### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [ ] Ran `make format`.
- [ ] Fixed errors surfaced by `make clippy`.

---

[Rendered](https://github.com/tock/tock/blob/ci-policy/doc/CodeReview.md#3-continuous-integration)

### CODE REVIEW

File: doc/CodeReview.md [hunk 1]
```diff
@@ -77,7 +78,87 @@ possible responses:
...
+To the extent reasonable, individual tests are broken into small atomic units.
+All actionable `make` recipes that run tests **must** be in `ci-job-*` rules.
+These perform individual, logical actions related only to testing.
+These rules **must not** perform any one-off or setup operations.
+No automated tooling should invoke job rules directly.
+If a CI check fails, developers should be able to run the failed `ci-job-*`
+locally, although in certain cases this may require installing supporting
+tooling.
+
+##### `ci-setup-*`
```
Comment (alistair23) [inline doc/CodeReview.md:126]: Will `make ci` run `ci-setup-*` and then `ci-job-*`

Comment (ppannuto) [reply doc/CodeReview.md:126]: Effectively, thought the setup would be more of:
```make
ci-job-format-check: ci-setup-rustfmt
     run rustfmt

ci-setup-rustfmt:
     install rustfmt

ci:  ci-job-format-check
```

File: doc/CodeReview.md [hunk 2]
```diff
@@ -77,7 +78,87 @@ possible responses:
...
+
+Generally, there are a series of fine-grained targets that do the actual tests,
+and then a meta layer of rules that are invoked depending on context.
+
+#### The short answer: `make ci`
+This is a meta-target that runs what Tock considers the "standard developer CI".
+This is the rule that should be run locally before submitting PRs.
+It runs the quicker jobs that catch the majority of small errors.
+
+
```
Comment (alistair23) [inline doc/CodeReview.md:110]: Double space here

File: doc/CodeReview.md [hunk 3]
```diff
@@ -77,7 +85,86 @@ possible responses:
...
+automatically clearing caches or other artifacts if needed.
+Setup targets **may** handle downgrades, but developers working on experimental
+branches may be required to handle these cases manually.
+Setup targets are permitted to expect "total ownership" of the directories they
+create and manage.
+
+Setup rules may vary between runner and local environments, as they may perform
+automatic and possibly invasive (e.g. apt install) operations on runners.
+
+When run locally, setup targets **may** prompt users during install events.
```
Comment (lschuermann) [inline doc/CodeReview.md:148]: Mhm. This implies that a `make ci` could potentially persistently and arbitrarily change my system by doing system-wide installs and or configuration of tools, right? I think that this _may_ be helpful for some people, however there are probably a lot of people (including me) who do not want this.

On a local (non CI-infrastructure) build, I don't expect any persistent changes of my system outside of the tock working directory, or _at least_ a polite question whether the program is allowed to do so. I therefore propose changing this to a `setup targets **must** prompt users prior to system-wide persistent changes`.

Comment (alistair23) [reply doc/CodeReview.md:148]: Agreed. I think in general we will avoid installing anything (as it's annoying and hard to do anyway).

**Comment by ppannuto** [comment]
Updated the setup language.

Changed `make ci` to `make prepush` to indicate its intent. @hudson-ayers that target won't be the same as what's currently defined as `make ci` [the currently defined `make ci` would be the `make ci-all` rule in this proposal]

File: doc/CodeReview.md [hunk 4]
```diff
@@ -77,7 +85,95 @@ possible responses:
...
+All CI is driven by `make` rules.
+
+Generally, there are a series of fine-grained targets that do the actual tests,
+and then a meta layer of rules that are invoked depending on context.
+
+#### The short answer: `make prepush`
+This is a meta-target that runs what Tock considers the "standard developer CI".
+This is the rule that should be run locally before submitting PRs.
+It runs the quicker jobs that catch the majority of small errors.
+Developers are encouraged to consider wiring this to the
```
Comment (hudson-ayers) [inline doc/CodeReview.md:116]:
```suggestion
Developers are encouraged to consider writing this to the
```

Comment (ppannuto) [reply doc/CodeReview.md:116]: I think I meant wiring?

Comment (hudson-ayers) [reply doc/CodeReview.md:116]: Oh, I see, I am dumb, sorry

**Pull request merged**